### PR TITLE
Change LazyListState to be scroll-aware

### PR DESCRIPTION
**Pull request by dnagler**
- Reduce the size of the preload window while actively scrolling
- Once scrolling is completed, optimize the preload window so that, for example, if the user is scrolling down, we load more items below than above
- Optimize the scroll window so that it doesn't evict rows overly eagerly

This addresses issue #2002, and is a second attempt at PR #2013. Easier to review commit-by-commit (the first commit is a direct cherry-pick).

---

- [x] `CHANGELOG.md`'s "Unreleased" section has been updated, if applicable.

### CODE REVIEW

File: redwood-lazylayout-compose/src/commonMain/kotlin/app/cash/redwood/lazylayout/compose/LazyListState.kt [hunk 1]
```diff
@@ -156,13 +158,38 @@ public open class LazyListState {
...
 
-    // TODO(dylan+jwilson): If we're contiguous with our previous loaded range,
-    //  don't rush to remove things from the previous range.
-    val begin = (firstIndex - preloadBeforeItemCount).coerceAtLeast(0)
-    val end = (lastIndex + preloadAfterItemCount).coerceAtMost(itemCount).coerceAtLeast(0)
+    var begin = (firstIndex - preloadBeforeItemCount).coerceAtLeast(0)
+    var end = (lastIndex + preloadAfterItemCount).coerceAtMost(itemCount).coerceAtLeast(0)
+
+    // On initial load, set lastIndex to the end of the loaded window.
+    if (lastIndex == 0) {
```
Comment (dnagler) [inline redwood-lazylayout-compose/src/commonMain/kotlin/app/cash/redwood/lazylayout/compose/LazyListState.kt:168]: @swankjesse Is this okay to do? I don't think it makes sense to keep (0, 0) as the range after the initial pass.

File: redwood-lazylayout-compose/src/commonMain/kotlin/app/cash/redwood/lazylayout/compose/LazyListState.kt [hunk 2]
```diff
@@ -106,4 +115,94 @@ public open class LazyListState {
     this.firstIndex = firstIndex
     this.lastIndex = lastIndex
   }
+
+  public fun loadRange(itemCount: Int): IntRange {
+    val preloadBeforeItemCount: Int
+    val preloadAfterItemCount: Int
+
+    when {
```
Comment (swankjesse) [inline redwood-lazylayout-compose/src/commonMain/kotlin/app/cash/redwood/lazylayout/compose/LazyListState.kt:135]: love it

File: redwood-lazylayout-compose/src/commonMain/kotlin/app/cash/redwood/lazylayout/compose/LazyListState.kt (hunk 1 above)
Comment (swankjesse) [reply redwood-lazylayout-compose/src/commonMain/kotlin/app/cash/redwood/lazylayout/compose/LazyListState.kt:168]: Seems reasonable!

File: redwood-lazylayout-compose/src/commonMain/kotlin/app/cash/redwood/lazylayout/compose/LazyListState.kt [hunk 3]
```diff
@@ -106,4 +115,94 @@ public open class LazyListState {
...
+
+    // If we're contiguous with the previous loaded range,
+    // don't rush to remove things from the previous range.
+    if (firstIndexFromPrevious1 != DEFAULT_SCROLL_INDEX &&
+      lastIndexFromPrevious1 != DEFAULT_SCROLL_INDEX
+    ) {
+      // Case one: Contiguous scroll down
+      // Previous: [firstIndexFromPrevious1...lastIndexFromPrevious1]
+      // Current requested range:           [begin.....................end]
+      // Solution: [firstIndexFromPrevious1............................end]
```
Comment (swankjesse) [inline redwood-lazylayout-compose/src/commonMain/kotlin/app/cash/redwood/lazylayout/compose/LazyListState.kt]: I LOVE THESE COMMENTS

**Review by swankjesse — APPROVED** [review]
So good

**Comment by underscoretang** [comment]
Please test this with money if you haven't before merging. I worry that we'll get some weird behavior with showing placeholders on hard scrolls again

**Comment by dnagler** [comment]
> Please test this with money if you haven't before merging. I worry that we'll get some weird behavior with showing placeholders on hard scrolls again

Absolutely, will do!

**Comment by dnagler** [comment]
- [x] Test end-to-end on Cash's Money and Activity tabs before merging

**Review by swankjesse — APPROVED** [review]

**Comment by dnagler** [comment]
> Please test this with money if you haven't before merging. I worry that we'll get some weird behavior with showing placeholders on hard scrolls again

@underscoretang Tested E2E on the Money tab--with an out-of-the-box `LazyListState`, we tend to evict some rows once they scroll offscreen (due to how tall some rows are), but by increasing the `scrollInProgressPreloadItemCount` to 10 in the Treehouse layer, the entire list stays rendered as expected.

File: redwood-lazylayout-compose/src/commonTest/kotlin/app/cash/redwood/lazylayout/compose/LazyListTest.kt [hunk 4]
```diff
@@ -101,12 +101,12 @@ class LazyListTest {
 
       with(awaitSnapshot()) {
         val lazyList = single() as LazyListValue
-        assertThat(lazyList.itemsBefore).isEqualTo(35)
-        assertThat(lazyList.itemsAfter).isEqualTo(25)
+        assertThat(lazyList.itemsBefore).isEqualTo(50 - 5)
+        assertThat(lazyList.itemsAfter).isEqualTo(100 - (60 + 20))
```
Comment (swankjesse) [inline redwood-lazylayout-compose/src/commonTest/kotlin/app/cash/redwood/lazylayout/compose/LazyListTest.kt:105]: nice

**Review by swankjesse — APPROVED** [review]

**Pull request merged**